### DISCUSSION TITLE
hy and hy-mode.el start-up behavior

### DISCUSSION THREAD

**Discussion by prairie-guy** [hylang/hy#2057] [Q&A] [answered]
I’m trying to configure `hy-mode` such that `hy-repl` in emacs acts like: 
`hy --repl-output-fn=hy.contrib.hy-repr.hy-repr`

Within my `config.el` file:
`(setq hy-shell--interpreter-args 'nil)` sets `repl-spy` to`False` as expected.
`(setq hy-shell--interpreter-args '("--repl-output-fn=hy.contrib.hy-repr.hy-repr”)` returns the following error on starting up error in `hy-repl`

Error: `hy: unrecognized option: --repl-output-fn\`

Is this an issue with mangling or improper use of delimiters?

Also, is there a default start-up file that `hy` uses or is the best way to set the environmental variable `HYSTARTUP`

If I am reading the documentation properly, it does look like I can set `repl-output-fn` at start-up. In a perfect world, I would like the set a global default start-up behavior, but still be able to override it within `config.el`.

Finally, is it possible to dynamically reset `hy-repl `behavior from within a running repl? I’ve tried executing “(setv repl-spy True)”

Appreciate the help. Looking to really start playing around with the language once I understand how to use it properly within emacs.

Thanks

**Comment by allison-casey** [comment, marked as answer]
you can set repl output fn with the following in your emacs configs
`(setq hy-shell--interpreter-args '("--repl-output-fn" "hy.contrib.pprint.pformat"))` hy-repr or pformat are good choices here.

I don't believe you can change the spy behavior while the repl is running. Pretty sure that's a limitation of python's codeop classes that the repl is built on top of. You've made me realize though that if you have a HYSTARTUP file that sets the spy param it won't get overridden if yoiu manually pass --spy. i'll open a PR to fix that

**Reply by prairie-guy** [reply]
That worked. I can now configure emacs hy-repl behavior. Thanks

**Comment by prairie-guy** [comment]
I am clearly missing something in trying to set up the HYSTARTUP file. I am using 0.20.0

I have added a test `startup.hy` file and added `export HYSTARTUP=/home/cdaniels/bin/startup.hy` to `.bashrc`. As you can see, if I load `startup.hy` with `hy -i`  from the command line, it works as expected. However, `startup.HY doesn’t appear to load when I just call `hy`. Furthermore, the correct environmental variable can be seen from within `hy` itself. (I looked through the source code to see if somehow the environmental variable differed from the docs, but in both cases, HYSTARTUP appears.)

Any advice appreciated.

Thanks in advance.

```
$ cat /home/cdaniels/bin/startup.hy
(print "HELLO")

$ hy -i /home/cdaniels/bin/startup.hy
HELLO
hy 0.20.0 using CPython(default) 3.8.5 on Linux
=> 
now exiting HyREPL...

$ hy 
hy 0.20.0 using CPython(default) 3.8.5 on Linux
=> (import os)
=> (os.environ.get "HYSTARTUP")
'/home/cdaniels/bin/startup.hy'
=> 
```

**Reply by allison-casey** [reply]
`HYSTARTUP` files were only added in `1.0a1` and wont be available on `0.20`. when looking at the docs make sure you are looking at the correct version. If you're looking to use supporting libs like jedhy, the appropriate docs would be [stable](https://docs.hylang.org/en/stable/) which you'll see doesn't make reference to `HYSTARTUP`. There's a panel at the bottom of the navigation panel on the docs that'll let you switch which versions of the docs you're running. Also check out [NEWS](https://github.com/hylang/hy/blob/master/NEWS.rst) for info about release features and changes

**Reply by prairie-guy** [reply]
OK, that would explain it! I cloned `1.0a1`, which is where I found what appeared to be functioning code (including tests) for `HYSTARTUP`. I hadn’t considered that it was new feature. 

At some point, would it be useful if I were to summarize my experiences in a more detailed installation guide? (Linux and OS X only.) If so, I would probably need to bug you with more questions . . . Or maybe I should wait until `1.0a1` is more stable.

I did read your discussion of the `Electric Boogaloo` and am thrilled that development is continuing. I really like being able to code in Clojure style, but having access to the Python ecosystem. (I also find Clojure too strict regarding immutability. Sometimes, you just want to update an array in place without having to jump through hoops.)

BTW, the cheatsheet is great! Very useful for a beginner. Comparing it to the Clojure cheatsheet really shows the influence of Clojure in the design of `HY`. Are you looking for PRs on other core functions for `Hy`? Not sure I could do anything too complicated, but for example, I have always wanted a `reverse` function in Python similar to the one in Clojure.

To get up to speed on the language, would you recommend `1.0a1` (and forego `jedhy` for now) or focus on `0.20`. To be clear, I’m just learning, I’m not planning to do anything more for now.

I really do appreciate you taking the time to respond to my newb requests.

(Also, on an iPadPro, the doc navigation switch doesn’t appear to work in Safari or Chrome, but does in Firefox)

**Reply by allison-casey** [reply]
Yes! That is the plan. I need to get a PR or two in and then i'm going to completely rewrite the docs and this has been great to figure out what we need to focus on, so thank you! 

As for core functions. We are slimming down the Hy library to remove functions that are better available elsewhere (such as the wonderful `toolz` lib). Everything else will be moved to a new hy stdlib repo with easier pull request requirements, so be on the lookout for that in the coming months if you want to contribute on that front. (btw, to reverse a sequence in python there's a builtin `reversed` method)

If you're not dependent on legacy libraries that only support `0.20` i'd recommend going with the alphas. You'll be better prepared for the breaking changes that will be coming in `1.0` (and you get some cool features like `HYSTARTUP`) and if you find any bugs in them we can support you better. 

Hmm noted, weird that safari and chrome to work in that case. Thanks!

**Reply by prairie-guy** [reply]
Alison - Looking forward to the new docs. When you do update them, I will read them wearing my first-time-user hat. (For me, a good cheatsheet and lots of examples are the best way to figure something out.)

`toolz` does look like a great library. It make sense to use it. I would hope that you would continue to provide most of the Clojure type functionality “out of the box” as part of the core language. Functional languages are finally becoming more popular. It seems many programmers are wanting to see what functional programming is all about. Given the large Python community, Hy would seem to be ideally situated to on-board many of those wanting to give FP a try. (Certainly more so than Clojure itself, Scala or Haskell.) I’m biased, but Clojure is an amazing language, one well worth emulating. In order to help new users appreciate the benefits of FP, it would seem that a well curated version of Hy with “batteries included” would be ideal. (Just my two-cents)

Would you also expect to promote `let` and `seq` and some of the other extras into the core language? 

I have just purged `0.20` and will start using `1.0a1`. Instead of running `inferior-hy-mode`, maybe I will just use `vterm` running `hy` instead. Do you mind if I ask what your set up is?

Thanks

**Reply by allison-casey** [reply]
I appreciate that! Feedback is always welcome

There is still discussion to be had about what exactly the new common library will look like. Will it import `toolz` methods along with the more clojure-like methods? What about python's `itertoolz` lib or by extension `more-itertoolz` (a fantastic lib btw)? All of which will be discussed in the coming alphas as we trim down the core language repo and move non critical methods over to the new repo. Hy will still come with a base set of function methods and macros that a clojurian would expect, just not the massive standard library you might expect coming from clojure. The reasoning for this is that the more stuff that gets packed into the hy compiler repo, the harder it is to focus on developing the language (especially for such a small team like us). Moving the extra fat and quality of life methods over to a separate repo will give us space to focus on the nitty gritty and lower the barrier to entry for people to contribute bugfixes, methods, and macros to the standard library without having to sift through core compiler code. The documentation will reference and recommend the stdlib and show how to use it in detail. 

I wouldn't expect `let` or `seq` to be promoted to core in `1.0`. The current focus is stabilizing the language syntax and api so that people can really start building with Hy. Maybe in a subsequent release once their stability and implementations are shown to be stable, but that will need to be discussed in its own issue down the road. 

I actually use `hy-mode` in emacs for all my programming in Hy. The only configuration I do is set the following.
```lisp
(setq hy-shell--interpreter-args '("--repl-output-fn" "hy.contrib.pprint.pformat")
      hy-jedhy--enable? nil)
```
I have some other stuff tailored to my specific configuration, but that's all I really do. I haven't found losing `jedhy` to be *that* bad of a tradeoff. `jedhy` mostly does runtime introspection with `dir` which i do myself at the repl
